core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -105,16 +105,31 @@ class ReplicaFetcherThread(name:String,
       leaderEndOffset
     } else {
       /**
-       * The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
+       * If the leader's log end offset is greater than follower's log end offset, there are two possibilities:
+       * 1. The follower could have been down for a long time and when it starts up, its end offset could be smaller than the leader's
        * start offset because the leader has deleted old logs (log.logEndOffset < leaderStartOffset).
+       * 2. When leader migration occurs, there is a slight chance that the old leader's high watermark is greater than
+       * the new leader's log end offset, so an OffsetOutOfRangeException was thrown. After that some more messages are
+       * produced to the leader. When the follower was trying to handle the exception and query the log end offset of
```

Comment:
@gwenshap Sorry for late reply... I somehow missed email from your followup comments...
I will put a more detailed sequence there but this could be a bit verbose. And people probably still need to read more code to understand what happened.